drop-boxes/register-raw-and-mzML/register-raw-and-mzml.py

```diff
@@ -0,0 +1,269 @@
+SAMPLE_TRACKER = SampleTracker.createLocationIndependentSampleTracker(SERVICE_REGISTRY_URL, SERVICE_CREDENTIALS)
+
+# *Q[Project Code]^4[Sample No.]^3[Sample Type][Checksum]*.*
+barcode_pattern = re.compile('Q[a-zA-Z0-9]{4}[0-9]{3}[A-Z][a-zA-Z0-9]')
+ms_pattern = re.compile('MS[0-9]*Q[A-Z0-9]{4}[0-9]{3}[A-Z][A-Z0-9]')
+ms_prefix_pattern = re.compile('MS[0-9]*')
+
+VENDOR_FORMAT_EXTENSIONS = {'.raw':'RAW_THERMO', '.d':'D_BRUKER','.wiff':'WIFF_SCIEX'}
+WATERS_FORMAT = "RAW_WATERS"
+
+def extract_barcode(filename):
+    """Extract valid barcodes from the filename.
+
+    Return project_id, experiment_id and the whole barcode.
+    If no barcode was found, raise a ValueError. If the file
+    contains a qbic barcode with an invalid checksum, raise
+    a RuntimeError.
+
+    TODO rewrite
+    """
+    try:
+        code = barcode[0:-1]
+        return checksum.checksum(code) == barcode[-1]
+    except:
+        return True
+
+def parse_timestamp_easy(mzml_path):
+    with open(mzml_path, 'r') as mzml:
+        time = None
+        for line in mzml:
+            if "<run id=" in line:
+                for token in line.split(" "):
+                    if "startTimeStamp=" in token:
+                        xsdDateTime = token.split('"')[1]
+                        time = datetime.datetime.strptime(xsdDateTime, '%Y-%m-%dT%H:%M:%SZ').strftime('%Y-%m-%d %H:%M:%S')
+                break
+        return time
+
+def parse_instrument_accession(mzml_path):
+    with open(mzml_path, 'r') as mzml:
+        accession = None
+        out = True
+        for line in mzml:
+            if "<instrumentConfigurationList" in line or 'id="CommonInstrumentParams">' in line:
+                out = False
+            if "</referenceableParamGroup>" in line or "</instrumentConfiguration>" in line:
+                out = True
+            if not out and '<cvParam cvRef="MS"' in line:
+                line = line.split(" ")
+                for token in line:
+                    if "accession=" in token:
+                        accession = token.split('"')[1]
+                break
+        print "accession for "+mzml_path+": "+accession
+        return accession
+
+def parse_timestamp_from_mzml(mzml_path):
+    schema = '{http://psi.hupo.org/ms/mzml}'
+    for event, element in xml.etree.ElementTree.iterparse(mzml_path):
+        if element.tag == schema+'run':
+            xsdDateTime = element.get('startTimeStamp')
+            element.clear()
+            break
+        element.clear() # remove unused xml elements
+    time = None
+    try:
+        time = datetime.datetime.strptime(xsdDateTime, '%Y-%m-%dT%H:%M:%SZ').strftime('%Y-%m-%d %H:%M:%S')
+    except TypeError:
+        print "no startTimeStamp found"
+    return time
+
+def createSimilarMSExperiment(tr, space, project, existing):
+    numberOfExperiments = len(existing)
+    newExpID = '/' + space + '/' + project + '/' + project + 'E' +str(numberOfExperiments)
```

Comment:
Why is the project Identifier two times in this ID?

Reply:
it's the project code, each experiment starts with the project code

Reply:
Ah that makes sense thanks for the explanation. The readability could be improved by replacing this with applying   [string.format ](https://www.w3schools.com/python/ref_string_format.asp) instead. 
However this is only nitpicking :)

Reply:
good idea, changed